src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -207,51 +207,104 @@ private void SignalCompletion()
+                    ExecutionContext.Restore(currentContext);
                 }
                 else
                 {
-                    InvokeContinuation();
+                    // Running inline may throw; capture the edi if it does as we changed the ExecutionContext,
+                    // so need to restore it back before propagating the throw.
+                    ExceptionDispatchInfo? edi = InvokeInlineContinuation();
+                    // Restore the current ExecutionContext.
+                    ExecutionContext.Restore(currentContext);
+                    // Now rethrow the exception; if there is one.
+                    edi?.Throw();
                 }
+
+                return;
             }
+
+            InvokeSchedulerContinuation();
+            // Restore the current ExecutionContext.
+            ExecutionContext.Restore(currentContext);
         }
 
         /// <summary>
-        /// Invokes the continuation with the appropriate captured context / scheduler.
-        /// This assumes that if <see cref="_executionContext"/> is not null we're already
+        /// Invokes the continuation inline and captures any exception thrown.
+        /// This assumes that if <see cref="_continuation"/> is not null we're already
         /// running within that <see cref="ExecutionContext"/>.
         /// </summary>
-        private void InvokeContinuation()
+        private ExceptionDispatchInfo? InvokeInlineContinuation()
         {
             Debug.Assert(_continuation != null);
+            Debug.Assert(_capturedContext == null);
+            Debug.Assert(!RunContinuationsAsynchronously);
 
-            switch (_capturedContext)
+            ExceptionDispatchInfo? edi = null;
+            SynchronizationContext? syncContext = SynchronizationContext.Current;
+            try
             {
-                case null:
-                    if (RunContinuationsAsynchronously)
-                    {
-                        if (_executionContext != null)
-                        {
-                            ThreadPool.QueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
-                        }
-                        else
-                        {
-                            ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
-                        }
-                    }
-                    else
-                    {
-                        _continuation(_continuationState);
-                    }
-                    break;
+                _continuation(_continuationState);
+            }
+            catch (Exception ex)
+            {
+                // Note: we have a "catch" rather than a "finally" because we want
+                // to stop the first pass of EH here.  That way we can restore the previous
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
```

Comment:
If the caller uses try/finally to restore the EC, can this catch block and return value be removed to have the exception propagate up natuarally?

Reply:
Is just following same pattern as `ExecutionContext.Run` (also uses same comment); not entirely sure what it means :)

Reply:
And it could totally be wrong... Not sure. CC @janvorli

Reply:
Unresolving for now, still curious if the catch block / return value are necessary, depending on @janvorli's insight, but no need to block on this

Reply:
I see that the return value is not relevant anymore after the latest changes, now just wondering if the catch block is necessary. Also should the rethrow of `edi` be in the finally block?

Reply:
From the EH point of view, the code in finally is executed on the 2nd pass while the exception filters upto the place where the exception is going to be caught are executed on the first pass. 
So filters will always run before finallys.

Reply:
Thanks @janvorli, then I guess the catch-all is necessary

Reply:
> Also should the rethrow of edi be in the finally block?

On second thought it doesn't seem like it would make a difference

Reply:
@janvorli does that mean that:

```c#
try { ... }
catch(Exception)
{
    // code
    throw;
}
```

Is fundamentally different in behavior in the presence of exception filters than:

```
try { ... }
finally
{
    // code
}
```

? If so, why should it be different?

Reply:
Edited inline above to hopefully make them code-wise equivalent

Reply:
Yes, it is. In the catch case above, the code in the catch is executed before any possible filters in the callers of this code. In the first pass of EH, the catch is found, so we scan no more and start the second pass. At the end of the second pass, the catch handler is invoked, the "code" executed. Then we rethrow the exception and a new exception handling starts, this time invoking possible filters in the callers of this code in the first pass.

In the finally case, the code is executed after the filters. The first pass of the EH is looking for a handler, so it executes the filters in the callers of this code to see if they can handle the exception. When a handler is found, then second pass starts, it calls the code in the finally and ultimately executes the handler it has found.

Reply:
> The first pass of the EH is looking for a handler, so it executes the filters in the callers of this code

Why is the finally block not considered a handler? Naturally the code in it would have to occur before any caller catch blocks. So why are caller exception filters considered before running the finally similarly to the first case?

Reply:
Handler that we look up in the first pass is a catch where the exception propagation stops and the exception is handled. In the second pass the stack is cleaned up and unwound. finally doesn't do any handling, it is just kind of cleanup on the second pass (considering the finally called exceptionally)

Reply:
If finally were to be considered a handler similarly to the first case, then would it be incorrect to unwind the stack up to the finally block (similarly to first case catch-all/rethrow), before checking exception filters? I'm still curious why it's necessary for exception filters to run before finally's.

Reply:
The exception filters are a way to determine where an exception is to be handled. When they return true, it means that the related catch is the handler. The other way is regular catch without filter that checks if the current exception type matches the one specified in the catch. This happens on the first pass. Only after the catch where the exception will be handled is found, we can start the 2nd pass where the finallys run and the catch handler is executed as the last thing.

I am not sure I have fully understood the question. I wonder if you were rather asking why finallys are not executed on the first pass?

Reply:
I'm wondering why upon encountering a finally we don't switch to second pass, similarly to a catch handler that satisfies the exception filter (like a catch with no exception filter that satisfies the exception type)

Reply:
In essence the difference in behavior between the above two cases seems unintuitive to me, and a result of implementation detail. I'm wondering why such a difference should exist.

Reply:
The exception handling way is described by the ECMA-335 spec, chapter "I.12.4.2.5 Overview of exception handling". It explicitly says that first, the catch handler is found and the finallys are executed after that.
It basically mimics Windows SEH behavior.

Reply:
Ah ok thanks @janvorli